### PR TITLE
Arcargo: Vendor Cargo and Vending Machine Update

### PR DESCRIPTION
**Pull request by ArcaneMusic**
Another one.

## About The Pull Request
This pull request originally had a design doc that @Fikou and I worked on, but that was never really polished up for publishing quality so I'll forgo it for now and be as descript as possible here.

### Core changes
- This pull request adds a new NTOS app to the game, the restock tracker. The restock tracker shows a comprehensive list of vending machines across the station, as long as there is a need for that vending machine to get restocked.
    - This has also been pre-installed into the cargo data disks. (`/obj/item/computer_disk/quartermaster`)
- Vending machines now store a total of 20% of the cost of any purchase made within themselves into a small pool of cash. This only applies to premium and normal purchases, not to contraband, as they're technically not sanctioned by the company.
- The restock tracker app will also track which vending machines have the most credits stored internally inside them.
- By refilling a vending machine, the stored credits within are paid out to any crewmember who goes and restocks the station, while also paying out *half that amount to the cargo budget*, serving as a basic but otherwise easy tertiary money making method on the same level of complexity as doing bounties, with the added benefit of actually helping to assist the station for jobs like... assistant.

image

### Break Stuff
- Anyway, when you try and smash a vending machine open with a melee weapon of choice, it can now pay out 50 credits at a time as a way to make money at zero risk to yourself.
- ~~Except for the horrible risk to yourself.~~
image

### Cargo Specific Changes
- Restock units may now be sold for a small profit as well, to incentivize cargo to keep the station stocked further.
- The `STATION_TRAIT_VENDING_SHORTAGE` trait will now add a small amount of existing credits into the vending machines on station, to incentivize cargo to fix the issue during the round and not just push for an early shuttle call. Or, more accurately, provide the crew with a money making scheme to engage better with the station trait as it stands.

### This also refactors behavior on vending machines
- This pull request also finally changes it so that vending machines now use the payment component, which as a consequence allows for the following improvements:
    * Vending machines may now pull from physical credits on your person, not just requiring you to have money on your ID card.
    * Vending machines may also use credits being pulled by the player interacting with the vending machine, allowing for handless mobs to be able to purchase items from a vending machine.
    * Finally makes the "use-for-everything buying things component" used by the most utilized component of the in-game economy, to reduce the quantity of unique implementations of purchasing things in the code.
- Existing vending specific checks are retained on before handing off behavior to the payment component, for behavior such as purchasing cigarettes/alcohol under the age of 18/21.

Notes:
- Vending machines will lose their internal credits stored when deconstructed, as a security measure.
- Vending machines will now show the total amount of credits that a mob has on their person, combining physical credits as well as credits held in their ID card to accurately portray their total wealth across the mob in question. 

## Why It's Good For The Game

First off, this is largely an excuse to move vending machine behavior over to the payment component for the purposes to less code copy-paste, and to try and make the implementation more wide-spread.

Second, this implements a new tertiary economy method to the game, in the same design space as bounties, which serve as common methods of making money without necessarily being specific to their job in question, with the primary goal of providing small amounts of work to the crew and a basic interaction with the economy system.

Additionally, it gives cargo more things they can do to assist the station, and a way to know which parts of the station need support as a result.
It improves the interaction between the vending shortage station trait as well, making it a challenge with depth as opposed to a more oppressive round change that players would rather reroll the game over.

Additionally, this makes a few price tweaks to vending restock modules as well to help incentivize buying some of the more minor restock kits, and a few select bumps on restocks that cover wide enough territory to necessitate fewer restocks.

## Changelog

:cl:
add: Vending machines now track how many credits have been spent on them, and when restocked will pay out that saved portion to the restocker, with a 50% match going to the cargo department.
add: Adds the restock tracker app, an NTOS app that tracks how well stocked the station's vending machine units are at a glance as well as how much is contained in each.
refactor: Vending machines now use the payment component for money handling behavior, meaning it will now accept held or pulled coins/cash/credits
qol: Attacking vending machines can drop a portion of it's stored credits, at the usual expected danger.
balance: Tweaked the cost of various restock modules up and down.
qol: Restock modules can now be sold for 50 credits.
/:cl:

### CODE REVIEW

**Comment by Fikou** [comment]
we're so back

File: code/modules/vending/restock_tracking/restock_tracker.dm [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#define LOW_STOCK_THRESHOLD 75
+
+/**
+ * The restock tracker computer keeps tabs on which vending machines on station are running low on stock.
+ * It can be used by cargo technicians to quickly find out which machines need to be restocked, and rewards crew
+ */
+
+/obj/machinery/computer/restock_tracker
```
Comment (Fikou) [inline code/modules/vending/restock_tracking/restock_tracker.dm:8]: i remember us talking about making this an ntos app, anything you ran into that prevented that?

**Comment by Iajret** [comment]
> Vending machines will lose their internal credits stored when deconstructed, as a security measure.

so there IS a way to restock vendors without deconstructing them first?

**Comment by Admiral-Gigaton** [comment]
Hey, not to add more to your work, but could you also patch the issue of chaplains vendor deleting the armor beacon because of the vendor shortage crate?

File: code/modules/vending/restock_tracking/restock_tracker.dm (hunk 1 above)
Comment (ArcaneMusic) [reply code/modules/vending/restock_tracking/restock_tracker.dm:8]: My poor memory

**Comment by DATA-xPUNGED** [comment]
> Needs to not reward credit for refilling a free vending machine, to prevent exploits using offstation vendors (empty vendor, refill for profit, repeat). A variable to exclude vendor types from the reward system regardless of free status would be welcome too, for syndicate vendors.

My understanding is that the credits in the machine come from the person that buys stuff from them, so if the machine is free, you won't get any credits. Limiting the tracker to station Z-level would be ideal though

**Comment by ArcaneMusic** [comment]
> > Needs to not reward credit for refilling a free vending machine, to prevent exploits using offstation vendors (empty vendor, refill for profit, repeat). A variable to exclude vendor types from the reward system regardless of free status would be welcome too, for syndicate vendors.
> 
> My understanding is that the credits in the machine come from the person that buys stuff from them, so if the machine is free, you won't get any credits. Limiting the tracker to station Z-level would be ideal though

This is a correct interpetation, it's 20% of the amount spent by the player, as well as after discounts like same-department discount.

The tracker to the best of my understanding *should* be only on station Z, as the global list containing relevant vending machines should be limited to Z levels that are on station, but if not it should be simple to tweak.

**Comment by ATH1909** [comment]
> By refilling a vending machine, the stored credits within are paid out to any crewmember who goes and restocks the station, while also paying out half that amount to the cargo budget

lore-wise, is the money being sent to the cargo budget from centcom (as a reward for doing their jobs)?

**Comment by ArcaneMusic** [comment]
> > By refilling a vending machine, the stored credits within are paid out to any crewmember who goes and restocks the station, while also paying out half that amount to the cargo budget
> 
> lore-wise, is the money being sent to the cargo budget from centcom (as a reward for doing their jobs)?

I'm going to go with either Nanotrasen, or the company who owns the vending machine as part of a contract with nanotrasen

**Comment by ArcaneMusic** [comment]
Works as a NTOS app now as well as a dedicated console as per fikou's request.

File: code/__HELPERS/global_lists.dm [hunk 2]
```diff
@@ -258,3 +258,10 @@ GLOBAL_LIST_INIT(WALLITEMS_EXTERIOR, typecacheof(list(
 	/obj/structure/camera_assembly,
 	/obj/structure/light_construct,
 )))
+
+/// A static typecache of all the money-based items that can be actively used as currency.
+GLOBAL_LIST_INIT(allowed_money, typecacheof(list(
+		/obj/item/stack/spacecash,
+		/obj/item/holochip,
+		/obj/item/coin
```
Comment (MrMelbert) [inline code/__HELPERS/global_lists.dm:266]:
```suggestion
	/obj/item/coin,
	/obj/item/holochip,
	/obj/item/stack/spacecash,
```
ABC order, trailing comma, and one less indentation

File: code/modules/vending/_vending.dm [hunk 3]
```diff
@@ -655,7 +690,11 @@
 				// instantiate canister if needed
 				var/transferred = restock(canister)
 				if(transferred)
-					to_chat(user, span_notice("You loaded [transferred] items in [src]."))
+					to_chat(user, span_notice("You loaded [transferred] items in [src][credits_contained > 0 ? ", and are rewarded [credits_contained] credits." : "."]"))
+					var/datum/bank_account/cargo_account = SSeconomy.get_dep_account(ACCOUNT_CAR)
+					cargo_account.adjust_money(round(credits_contained * 0.5), "Vending: Restock")
+					var/obj/item/holochip/payday = new /obj/item/holochip(src, credits_contained)
```
Comment (MrMelbert) [inline code/modules/vending/_vending.dm:696]:
```suggestion
					var/obj/item/holochip/payday = new(src, credits_contained)
```

File: code/modules/vending/_vending.dm [hunk 4]
```diff
@@ -1396,26 +1425,26 @@
...
+/obj/machinery/vending/proc/proceed_payment(obj/item/card/id/paying_id_card, datum/data/vending_product/product_to_vend, price_to_use, mob/living/alive_user)
 	var/datum/bank_account/account = paying_id_card.registered_account
 	if(account.account_job && account.account_job.paycheck_department == payment_department)
 		price_to_use = max(round(price_to_use * DEPARTMENT_DISCOUNT), 1) //No longer free, but signifigantly cheaper.
 	if(coin_records.Find(product_to_vend) || hidden_records.Find(product_to_vend))
 		price_to_use = product_to_vend.custom_premium_price ? product_to_vend.custom_premium_price : extra_price
 	if(LAZYLEN(product_to_vend.returned_products))
 		price_to_use = 0 //returned items are free
-	if(price_to_use && !account.adjust_money(-price_to_use, "Vending: [product_to_vend.name]"))
+	if(price_to_use && attempt_charge(src, alive_user, price_to_use) & COMPONENT_OBJ_CANCEL_CHARGE)
```
Comment (MrMelbert) [inline code/modules/vending/_vending.dm:1436]:
```suggestion
	if(price_to_use && (attempt_charge(src, alive_user, price_to_use) & COMPONENT_OBJ_CANCEL_CHARGE))
```
Be careful with bitflag operations, they might own you in byond if you're not careful
Also, is this proc safe if passed a `null` user?

File: code/modules/vending/_vending.dm [hunk 5]
```diff
@@ -1396,26 +1425,26 @@
  * product_to_vend - the product record of the item we're trying to vend
  * price_to_use - price of the item we're trying to vend
  */
-/obj/machinery/vending/proc/proceed_payment(obj/item/card/id/paying_id_card, datum/data/vending_product/product_to_vend, price_to_use)
+/obj/machinery/vending/proc/proceed_payment(obj/item/card/id/paying_id_card, datum/data/vending_product/product_to_vend, price_to_use, mob/living/alive_user)
```
Comment (MrMelbert) [inline code/modules/vending/_vending.dm:1428]: `alive_user` is kinda a weird name for this, should give it a more accurate name for what the user is doing here, like I assume it's the mob paying

Also don't forget to update the doc comment

Also also putting this at the end is very weird, should put it after `paying_id_card` IMO (quick glance says there's only one use of this proc and one override so shouldn't be painful)

File: code/modules/vending/_vending.dm [hunk 6]
```diff
@@ -1554,6 +1583,19 @@
...
+/obj/machinery/vending/proc/deploy_credits()
+	if(credits_contained <= 0)
+		return
+	var/obj/item/holochip/holochip
+	if(credits_contained > CREDITS_DUMP_THRESHOLD)
+		holochip = new /obj/item/holochip(loc, CREDITS_DUMP_THRESHOLD)
+		credits_contained -= CREDITS_DUMP_THRESHOLD
+	else if(credits_contained > 0)
+		holochip = new /obj/item/holochip(loc, round(credits_contained))
+		credits_contained = 0
```
Comment (MrMelbert) [inline code/modules/vending/_vending.dm:1596]: Uh, work smarter not harder?
```suggestion
	var/credits_to_remove = min(CREDITS_DUMP_THRESHOLD, round(credits_contained))
	var/obj/item/holochip/holochip = new(loc, credits_to_remove)
	credits_contained = max(0, credits_contained - credits_to_remove)
```

File: code/modules/vending/restock_tracking/restock_tracker.dm [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+		var/stock = vendor.total_loaded_stock()
+		var/max_stock = vendor.total_max_stock()
+		if((max_stock == 0 || (stock >= max_stock)) && vendor.credits_contained == 0)
+			continue
+		vending_list += list(list(
+			"name" = vendor.name,
+			"location" = get_area_name(vendor),
+			"credits" = vendor.credits_contained,
+			"percentage" = (stock / max_stock) * 100,
+			"id" = id_increment
```
Comment (MrMelbert) [inline code/modules/vending/restock_tracking/restock_tracker.dm:34]:
```suggestion
			"id" = id_increment,
```

File: code/modules/modular_computers/file_system/programs/restock_tracker.dm [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+	can_run_on_flags = PROGRAM_LAPTOP | PROGRAM_PDA
+	size = 4
+	program_icon = "cash-register"
+	tgui_id = "NtosRestock"
+
+/datum/computer_file/program/restock_tracker/ui_data()
+	var/list/data = list()
+	var/list/vending_list = list()
+	var/id_increment = 1
+	for(var/obj/machinery/vending/vendor in GLOB.vending_machines_to_restock)
```
Comment (MrMelbert) [inline code/modules/modular_computers/file_system/programs/restock_tracker.dm:17]: `as anything`?

File: code/modules/modular_computers/file_system/programs/restock_tracker.dm [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+		var/stock = vendor.total_loaded_stock()
+		var/max_stock = vendor.total_max_stock()
+		if((max_stock == 0 || (stock >= max_stock)) && vendor.credits_contained == 0)
+			continue
+		vending_list += list(list(
+			"name" = vendor.name,
+			"location" = get_area_name(vendor),
+			"credits" = vendor.credits_contained,
+			"percentage" = (stock / max_stock) * 100,
+			"id" = id_increment
```
Comment (MrMelbert) [inline code/modules/modular_computers/file_system/programs/restock_tracker.dm:27]:
```suggestion
			"id" = id_increment,
```

File: code/modules/vending/restock_tracking/restock_tracker.dm [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+	circuit = /obj/item/circuitboard/computer/restock
+
+/obj/machinery/computer/restock_tracker/ui_interact(mob/user, datum/tgui/ui)
+	. = ..()
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Restock", name)
+		ui.open()
+
+/obj/machinery/computer/restock_tracker/ui_data(mob/living/carbon/human/user)
```
Comment (MrMelbert) [inline code/modules/vending/restock_tracking/restock_tracker.dm:19]: I just noticed this is entirely copy pasted between the app and the computer. Why have the computer anymore? Just use a mod PC pre-loaded with the app?

File: code/modules/mob/living/living.dm [hunk 11]
```diff
@@ -2645,3 +2645,20 @@ GLOBAL_LIST_EMPTY(fire_appearances)
...
 		look_down()
+
+/**
+ * Totals the physical cash on the mob and returns the total.
+ */
+/mob/living/verb/tally_physical_credits()
+	//Here is all the possible non-ID payment methods.
+	var/list/counted_money = list()
+	var/physical_cash_total = 0
+	for(var/obj/item/credit in typecache_filter_list(get_all_contents(), GLOB.allowed_money)) //Coins, cash, and credits.
```
Comment (MrMelbert) [inline code/modules/mob/living/living.dm:2656]: `as anything`?

File: code/modules/research/techweb/all_nodes.dm [hunk 12]
```diff
@@ -1232,6 +1232,7 @@
 		"prisonmanage",
 		"secdata",
 		"vendor",
+		"restock_tracker",
```
Comment (MrMelbert) [inline code/modules/research/techweb/all_nodes.dm:1235]: ABC order

File: code/modules/vending/_vending.dm [hunk 13]
```diff
@@ -611,6 +624,28 @@
...
 
+/**
+ * Returns the total amount of items in the vending machine based on the product records and premium records, but not contraband
+ */
+/obj/machinery/vending/proc/total_loaded_stock()
+	var/total = 0
+	for(var/datum/data/vending_product/record as anything in product_records)
+		total += record.amount
+	for(var/datum/data/vending_product/record as anything in coin_records)
+		total += record.amount
```
Comment (MrMelbert) [inline code/modules/vending/_vending.dm:635]:
```suggestion
	for(var/datum/data/vending_product/record as anything in product_records + coin_records)
		total += record.amount
```
/shrug ?

File: code/modules/vending/_vending.dm [hunk 14]
```diff
@@ -611,6 +624,28 @@
...
+
+/**
+ * Returns the total amount of items in the vending machine based on the product records and premium records, but not contraband
+ */
+/obj/machinery/vending/proc/total_max_stock()
+	var/total_max = 0
+	for(var/datum/data/vending_product/record as anything in product_records)
+		total_max += record.max_amount
+	for(var/datum/data/vending_product/record as anything in coin_records)
+		total_max += record.max_amount
```
Comment (MrMelbert) [inline code/modules/vending/_vending.dm:646]:
```suggestion
	for(var/datum/data/vending_product/record as anything in product_records + coin_records)
		total_max += record.max_amount
```

File: code/modules/vending/_vending.dm (hunk 13 above)
Comment (ArcaneMusic) [reply code/modules/vending/_vending.dm:635]: Is that even legal? 👀

File: code/modules/vending/restock_tracking/restock_tracker.dm (hunk 10 above)
Comment (ArcaneMusic) [reply code/modules/vending/restock_tracking/restock_tracker.dm:19]: I lean towards soul, but...

Yeah, you're right...

**Comment by ArcaneMusic** [comment]
Updated and description has been tweaked on the move over to just a pure NTOS app

**Comment by kassori** [comment]
I hate this PR so much, I've played 4 shifts in the last week and the job vendors that are required to do my job are missing half their seeds.

**Pull request merged**